operator/pkg/model/ingestion/ingress.go

```diff
@@ -0,0 +1,197 @@
+			Hostname: "*",
+			Routes: []model.HTTPRoute{
+				{
+					Backends: []model.Backend{
+						backend,
+					},
+				}},
+			Port: 80,
+		}
+
+		l.Sources = model.AddSource(l.Sources, sourceResource)
+
+		insecureListenerMap["*"] = l
+	}
+
+	// Now, we range across the rules, adding them in as listeners.
+	for _, rule := range ing.Spec.Rules {
+
+		host := "*"
+
+		if rule.Host != "" {
+			host = rule.Host
+		}
+
+		l, ok := insecureListenerMap[host]
+		l.Port = 80
+		l.Sources = model.AddSource(l.Sources, sourceResource)
+		if !ok {
+			l.Name = fmt.Sprintf("ing-%s-%s-%s", ing.Name, ing.Namespace, host)
+		}
+
+		l.Hostname = host
+		for _, path := range rule.HTTP.Paths {
+
+			route := model.HTTPRoute{}
+
+			switch *path.PathType {
+			case slim_networkingv1.PathTypeExact:
+				route.PathMatch.Exact = path.Path
+			case slim_networkingv1.PathTypePrefix:
+				route.PathMatch.Prefix = path.Path
+			case slim_networkingv1.PathTypeImplementationSpecific:
+				route.PathMatch.Regex = path.Path
+			default:
+				// TODO(youngnick): Figure out what to do in this case.
+				continue
+			}
+
+			backend := model.Backend{
+				Name:      path.Backend.Service.Name,
+				Namespace: ing.Namespace,
+			}
+			if path.Backend.Service != nil {
+				backend.Port = &model.BackendPort{}
+				if path.Backend.Service.Port.Name != "" {
+					backend.Port.Name = path.Backend.Service.Port.Name
+				}
+				if path.Backend.Service.Port.Number != 0 {
+					backend.Port.Port = uint32(path.Backend.Service.Port.Number)
+				}
+			}
+			route.Backends = append(route.Backends, backend)
+			l.Routes = append(l.Routes, route)
+		}
+
+		insecureListenerMap[host] = l
+	}
+
+	secureListenerMap := make(map[string]model.HTTPListener)
+
+	// First, we check for TLS config, and set them up with Listeners to return.
+	for _, tlsConfig := range ing.Spec.TLS {
+
+		for _, host := range tlsConfig.Hosts {
+
+			var l model.HTTPListener
+
+			l, ok := secureListenerMap[host]
+			if !ok {
+				l = insecureListenerMap[host]
+			}
+
+			l.Sources = model.AddSource(l.Sources, sourceResource)
+			if tlsConfig.SecretName != "" {
+				l.TLS = &model.TLSSecret{
+					Name: tlsConfig.SecretName,
+					// Secret has to be in the same namespace as the Ingress.
+					Namespace: ing.Namespace,
+				}
+
+			}
+			l.Port = 443
+			l.Hostname = host
+			secureListenerMap[host] = l
+
+			defaultListener, ok := insecureListenerMap["*"]
+			if ok {
+				// A default listener already exists, each Host in TLSConfig.Hosts
+				// needs to have a Listener configured that's a copy of it.
+				if tlsConfig.SecretName != "" {
+					defaultListener.TLS = &model.TLSSecret{
+						Name: tlsConfig.SecretName,
+						// Secret has to be in the same namespace as the Ingress.
+						Namespace: ing.Namespace,
+					}
+				}
+				defaultListener.Hostname = host
+				defaultListener.Port = 443
+				secureListenerMap[host] = defaultListener
+
+			}
+		}
+
+	}
+
+	listenerSlice := []model.HTTPListener{}
```

Comment:
`listenerSlice := make([]model.HTTPListener{}, 0, len(insecureListenerMap)+len(secureListenerMap))`?

In Cilium we tend to do things like this to minimize allocations while handling these resources.